src/compression/mod.rs

```diff
@@ -408,16 +408,103 @@ fn mod_p(x: i32, y: i32) -> i32 {
 mod optimize_bytes {
 
     /// Integrate over all differences to the previous value in order to reconstruct sample values.
-    pub fn differences_to_samples(buffer: &mut [u8]){
-        for index in 1..buffer.len() {
-            buffer[index] = (buffer[index - 1] as i32 + buffer[index] as i32 - 128) as u8; // index unsafe but handled with care and unit-tested
+    pub fn differences_to_samples(buffer: &mut [u8]) {
+        // The naive implementation is very simple:
+        //
+        // for index in 1..buffer.len() {
+        //    buffer[index] = (buffer[index - 1] as i32 + buffer[index] as i32 - 128) as u8;
+        // }
+        //
+        // But we process elements in pairs to take advantage of instruction-level parallelism.
+        // When computations within a pair do not depend on each other, they can be processed in parallel.
+        // Since this function is responsible for a very large chunk of execution time,
+        // this tweak alone improves decoding performance of RLE images by 20%.
+        if let Some(first) = buffer.get(0) {
+            let mut previous = *first as i16;
+            for chunk in &mut buffer[1..].chunks_exact_mut(2) {
+                // no bounds checks here due to indices and chunk size being constant
+                let diff0 = chunk[0] as i16;
+                let diff1 = chunk[1] as i16;
+                // these two computations do not depend on each other, unlike in the naive version,
+                // so they can be executed by the CPU in parallel via instruction-level parallelism
+                let sample0 = (previous + diff0 - 128) as u8;
+                let sample1 = (previous + diff0 + diff1 - 128 * 2) as u8;
```

Comment:
very interesting approach, it surprises me that it's faster, as we're now doing the computation twice! how did you come up with this approach? is that a common technique?

Reply:
Avoiding data dependencies to leverage instruction-level parallelism is a common technique, yes.

I have tried doing more, and it's a very slight improvement for 3 elements on my machine and about the same on 4 elements. I felt doing 3 might penalize less sophisticated hardware like a Raspberry Pi and wouldn't be worth it.

After writing this, I checked what OpenEXR is doing here,, and it has an implementation using explicit SIMD for SSE. But it's a lot of instructions, and at a glance I'm not convinced that it's going to be any faster than this. Plus this library is safe code only, so no explicit SIMD until portable SIMD stabilizes.

This is essentially computing a prefix sum, which is a common and well-studied primitive (well, with a few tweaks - the +previous and -128 bits, but those are easy to vectorize). You can find literature on it if you're willing to dig around. It's probably going to be talking about either thread parallelism or explicit SIMD, though.

Reply:
I never doubted that this is the best approach for the library, I have also thought about adding explicit or implicit SIMD from the start, it's great that you took the time! :) I was just curious, don't worry :)

Reply:
Oh, I'm not convinced that this is the best possible approach for the library. But it's nearly 2x faster than the current one, so I took it.

Reply:
It's less code and less complexity than explicit SIMD, and it is safe. How is this not the best approach for exrs? :)

Reply:
Maybe there's a clever solution that relies on autovectorization that I'm not seeing. Or someone came up with a super clever algorithm that leverages instruction-level parallelism even more. Or a myriad of other possibilities.

Reply:
I'd consider that to be your approach:) not saying it will never be improved, but that it's the right direction in my point of view